### PR TITLE
DRY RUN: Add semantic release and eslint

### PR DESCRIPTION
**Pull request by zinabat**
## Context
Hooray consistency!
- Adds semantic release github action
- Adds Coveralls action
- Prepares package for npm release
- Adds `eslint` and `eslint-brightspace-config`
- Updates code to pass linting

## Quality
- When this PR is merged, semantic release should run and release a new major version because of the `BREAKING CHANGE:` commit message. Let's get this pup deployed to npm 😄

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -13,14 +13,20 @@ jobs:
       - uses: Brightspace/third-party-actions@actions/cache
         id: cache
         with:
-          path: '**/node_modules'
+          path: |
+            **/node_modules
+            **/coverage
```
Comment (dlockhart) [inline .github/workflows/ci.yml:18]: This is going to cache the contents of `**/coverage`, but the cache key is just a hash of the package-lock file... so not sure why bother caching it? Isn't it going to get blown away when you re-run the Coveralls step down below anyway?

File: .eslintrc [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+{
+	"extends": ["brightspace/node-config"],
+	"env": {
+		"browser": true
+	},
+	"parserOptions": {
+		"ecmaVersion": 6,
+		"sourceType": "module"
+	}
+}
```
Comment (dlockhart) [inline .eslintrc:11]: A comparable config is likely the Intl one:
https://github.com/BrightspaceUI/intl/blob/main/.eslintrc.json

I think at the very least you want the `browser-config` as the extension point.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        with:
+          path: |
+            **/node_modules
+            **/coverage
+          key: npm-${{ hashFiles('**/package-lock.json') }}
+      - name: Install
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: npm install
+      - name: Run Tests
+        if: steps.cache.outputs.cache-hit != 'true'
```
Comment (dlockhart) [inline .github/workflows/release.yml:29]: I think you still want to re-run the tests even if there was a cache hit, no? The cache only invalidates when the package-lock file changes, but lots of other things can change (like code changes) that'll make you want to re-run tests.

Alternately, if we assume that the `ci.yml` workflow is for testing (which is what we do in core), then you could skip the lint/test/coveralls stage in the release workflow.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        run: npm run test:unit && npm run report-coverage
+      - name: Coveralls
+        uses: Brightspace/third-party-actions@coverallsapp/github-action
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build
+        run: npm run build
+      - name: Semantic Release
+        uses: BrightspaceUI/actions/semantic-release@master
+        with:
```
Comment (dlockhart) [inline .github/workflows/release.yml:39]: You'll want to pass `DEFAULT_BRANCH: main` here as an option. I'm going to flip the default eventually so `main` is the default, but it's a big job.

File: .npmignore [hunk 5]
```diff
@@ -1,5 +1,4 @@
-node_modules
-coverage
-test
-.editorconfig
-.travis.yml
+node_modules
+coverage
+test
+.editorconfig
```
Comment (dlockhart) [inline .npmignore:4]: Ever since [reading this](https://medium.com/@jdxcode/for-the-love-of-god-dont-use-npmignore-f93c08909d8d), I've been a fan of _not_ using an `.npmignore` file and just using the `files` entry in `package.json` to explicitly list out the things you want included in the package.

File: README.md [hunk 6]
```diff
@@ -101,9 +100,39 @@ npm run test:unit # run unit tests with coverage
...
 
-[npm-url]: https://npmjs.org/package/return-points
-[npm-image]: https://badge.fury.io/js/return-points.png
-[ci-image]: https://travis-ci.org/Brightspace/return-points.svg?branch=master
-[ci-url]: https://travis-ci.org/Brightspace/return-points
-[coverage-image]: https://img.shields.io/coveralls/Brightspace/return-points.svg
-[coverage-url]: https://coveralls.io/r/Brightspace/return-points?branch=master
+## Versioning & Releasing
+
+> TL;DR: Commits prefixed with `fix:` and `feat:` will trigger patch and minor releases when merged to `master`. Read on for more details...
```
Comment (dlockhart) [inline README.md:105]:
```suggestion
> TL;DR: Commits prefixed with `fix:` and `feat:` will trigger patch and minor releases when merged to `main`. Read on for more details...
```

File: package.json [hunk 7]
```diff
@@ -1,33 +1,41 @@
 {
-  "name": "return-points",
+  "name": "@brightspace-ui/return-points",
   "version": "1.0.0",
```
Comment (dlockhart) [inline package.json:3]: Since the last known release is `0.0.2` I think you actually want this to match that. Then I'd recommend merging this with a "BREAKING CHANGE" release, which will automatically bump things up to `1.0.0`.

It doesn't matter a huge amount since we're moving the NPM location anyway, so that old "return-points" will be stuck on 0.0.2 forever anyway.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -13,14 +13,20 @@ jobs:
...
         run: npm run lint
       - name: Test
-        run: npm run test:unit
+        run: npm run test:unit && npm run report-coverage
       - name: Build
         run: npm run build
+      - name: Coveralls
+        uses: Brightspace/third-party-actions@coverallsapp/github-action
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (dlockhart) [inline .github/workflows/ci.yml:32]: Cool! I'm curious why the "Coveralls" user had to be given access to the repo? I would have expected the token here to be enough.

File: package.json [hunk 9]
```diff
@@ -1,33 +1,41 @@
...
   "description": "A set of APIs to get/set a URL to \"return to\" after launching into a workflow that may have multiple entry points",
   "main": "index.js",
   "type": "module",
+  "publishConfig": {
+    "access": "public"
+  },
+  "files": [
+    "src"
```
Comment (dlockhart) [inline package.json:11]: Should `index.js` be included here?

Also curious... if `index.js` is the rollup-generated UMD thing that people use when doing `require('whatever')`, how does it also work with the `import { this, that } from 'thing'` syntax?

File: .github/workflows/release.yml (hunk 3 above)
Comment (zinabat) [reply .github/workflows/release.yml:29]: ah, yes this is from when I was trying to set up CI to also trigger before the release workflow, but I couldn't figure out how to make release happen after CI.. I'll play with it and see if I can make it in line with core

File: .npmignore (hunk 5 above)
Comment (zinabat) [reply .npmignore:4]: fair - this is leftover from the original I think. Will remove

File: .github/workflows/ci.yml (hunk 8 above)
Comment (zinabat) [reply .github/workflows/ci.yml:32]: I was hoping to get the coveralls bot to comment on new PRs with coverage info - this was listed in their docs as a step

File: package.json (hunk 9 above)
Comment (zinabat) [reply package.json:11]: I thought I tested this file with both a module with import statements and a commonjs file with require syntax, but now I see that UMD is not supported in browsers. Darn. Let me revisit

Comment (dlockhart) [reply package.json:11]: Since this is a major bump AND a new NPM module publish name anyway, we could just go with ESM-only. For places that want to `require('')` the CommonJS way, they can continue to use the old version. I'd be totally fine with that, and that's how other libraries like our Intl package work.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (zinabat) [reply .github/workflows/ci.yml:18]: I was essentially trying before to trigger a workflow from a workflow, but couldn't get that working so am removing this

File: package.json (hunk 9 above)
Comment (zinabat) [reply package.json:11]: > Should `index.js` be included here?

Files declared as `main` are [included by default](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#files) - and if no main is decalred, index.js is assumed as default anyway.
I'm restructuring things to simply include the `dist` folder and compile both cjs and esm versions anyway so moot point, but I thought it was neat

File: .github/workflows/release.yml (hunk 4 above)
Comment (zinabat) [reply .github/workflows/release.yml:30]: Thanks, I knew I'd probably missed one

**Review by dlockhart — APPROVED** [review]

File: .eslintignore [hunk 10]
```diff
@@ -0,0 +1 @@
+index.js
```
[the remark this answers is not in the document]
Comment (zinabat) [reply .eslintignore:1]: I could have sworn I removed it, grr
